cpp/src/lance/encodings/binary.cc

```diff
@@ -41,12 +41,18 @@ Result<int64_t> VarBinaryEncoder::Write(const std::shared_ptr<::arrow::Array> da
   ARROW_ASSIGN_OR_RAISE(auto offsets_position, out_->Tell());
   offsets_builder_.Reset();
   assert(arr->length() > 0);
+  /// Reset the slice's first offset to zero.
+  auto first_offset = arr->value_offset(0);
   for (int64_t i = 0; i <= arr->length(); ++i) {
-    ARROW_RETURN_NOT_OK(offsets_builder_.Append(start_offset + arr->value_offset(i)));
+    ARROW_RETURN_NOT_OK(
+        offsets_builder_.Append(start_offset + arr->value_offset(i) - first_offset));
   }
   ARROW_RETURN_NOT_OK(offsets_builder_.Finish(&offsets_));
+  assert(offsets_->Value(0) == start_offset);
 
-  ARROW_RETURN_NOT_OK(out_->Write(offsets_->values()));
+  ARROW_RETURN_NOT_OK(
+      out_->Write(offsets_->raw_values(),
+                  (arr->length() + 1) * sizeof(typename ::arrow::TypeTraits<OffsetType>::CType)));
```

Comment:
What does this mean?

Reply:
`offsets_->values()` is the underneath buffer shared by potentially multiple arrays.

`raw_values()` is like `values() + buffer.offset` which point to the data that this exact array holds.

```
sizeof(typename ::arrow::TypeTraits<OffsetType>::CType)
```

so it basically write `(len + 1)` offsets into disk.